### PR TITLE
feat: add User Search

### LINKED ISSUE #1647: Feature: User searchbar
### Type of feature

🍕 Feature

### Current behavior

N/A

### Context
This is a searchbar to allow users to find users on the platform. This is a first iteration of what we want to build into a command palette, the idea is to start with only the ability to search users and in the future add repo and insights/lists page search, highlights and more.

### [Figma Design Link](https://www.figma.com/file/OpVX6WT7dmWqnwRuEvADMF/Design-Lab?type=design&node-id=10369-17266&mode=design&t=oE7SBjhBJB8ttP6D-4)
**Design Lab** file > **Design: Search** page

## Functional requirements
Users will be able to type a username and get results for other users on OpenSauced
- Results should contain users that have connected and haven't connected their Github accounts. If they are in our database they should appear
- Max amount of results displayed is 3. The logic behind this is that as we turn it into a command palette we will add more features like searching for repos, so we want to keep the amount of results of each search category to a low number as to not clutter the ui.
- Users can press both the ⌘K command and the  /  key to trigger the searchbar (both are standard patterns for searchbars)
- Users can navigate the searchbar results using only the up/down arrows and the enter key to select a user profile.

### Default state
Show the search bar in the nav bar, next to the user avatar.

image

### Active state
When active, a new component will show up on top of the UI and the background will be slightly blurred.
- If the user hasn't typed anything, then show a "Type 3 characters or more to start searching" message at the bottom. Message will be displayed until the user types 3 minimum characters.
image

### Exact Match
If user inputs a string that exactly matches a username, only show that username

image

### Partial Match
If user inputs a string that partially matches multiple usernames, show up to 3 results that match the search (if there's only 2 results that match, show only 2 of course). 

image

### Exact + Partial Matches
If the user types a username and there’s an exact match, but there’s also partial matches:
Show exact match on the top and then up to 2 partial match results --> 3 results total

image

### No match
image

## Component

The searchbar component that will go on the nav bar is a slight variation of the existing Search component [link](https://design-insights.netlify.app/?path=/docs/design-system-atoms-search--docs) with an added badge showing the shortcut to trigger the searchbar (⌘K). 

The active searchbar component is meant to have the same look, but bigger.

image

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by babblebey**
## Description

This Pull Request introduces a User Search feature to enhance the OpenSauced platform's search capabilities. 

### Changes Made

- **SearchDialog Component**: A new `SearchDialog` component has been added. This component is displayed when a user clicks the `SearchDialogTrigger` component, providing an intuitive way to initiate searches.

- **Keyboard Shortcuts**: implemented keyboard shortcuts for quick access to the search functionality. Pressing `Ctrl + K` opens the `SearchDialog`. And, the `Escape` key can be used to close the dialog. 

- **Keyboard Controls**: implemented controls to navigate search results using the `ArrowUp` and `ArrowDown` keys, which activate a cursor trackable by a `_cursorActive` class name within the `UserResultCard` list controlled by the `cursor` state. And pressing `Enter` when a result is selected will navigate to the associated user profile page.

```js
const handleKeyboardCtrl: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
  const resultsCount = searchResult?.data?.length || 0;

  // ArrowUp 🔼 
  if (resultsCount && e.key === "ArrowUp") {
    e.preventDefault();
    setCursor(cursor === 0 ? Math.min(resultsCount - 1, 9) : cursor - 1);
  };

  // ArrowDown 🔽 
  if (resultsCount && e.key === "ArrowDown") {
    e.preventDefault();
    setCursor(cursor === Math.min(resultsCount - 1, 9) ? 0 : cursor + 1);
  };

  // Enter ↩️ 
  if (resultsCount && e.key === "Enter") {
    e.preventDefault();
    if (document.querySelector("._cursorActive")) {
      const user = document.querySelector("._cursorActive") as HTMLAnchorElement;
      router.push(user.href);
      setOpenSearch(false);
    }
  }
};
```

- **Search Info and Error Components**: implemented the `SearchInfo` and `SearchError` components according to the feature design requirement. The `SearchInfo` component provides guidance on the minimum characters required to initiate a search, while the `SearchError` component displays when no results are found, keeping users informed.

image

image

- **Search Loading Component**: implemented a `SearchLoading` component to manage transitions between loading and ready states.

image

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Fixes #1647 

## Mobile & Desktop Screenshots/Recordings

[screencast-localhost_3000-2023.09.21-00_17_00.webm](https://github.com/open-sauced/app/assets/25631971/06522ae7-38fe-4f59-ba79-6f65412a681d)

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?

NA

### CODE REVIEW

**Comment by OgDev-01** [comment]
Hey @babblebey, The API for this is ready [here](https://beta.api.opensauced.pizza/#/User%20service/getUsersByFilter) FYI 👍

**Comment by babblebey** [comment]
> Hey babblebey, The API for this is ready [here](https://beta.api.opensauced.pizza/#/User%20service/getUsersByFilter) FYI 👍 

Great stuff, thanks @OgDev-01

**Comment by isabensusan** [comment]
@babblebey  Can we do something like this where we only show the search icon?
image

**Comment by bdougie** [comment]
How is this coming along @babblebey. Are you blocked?

**Comment by babblebey** [comment]
Oh my apologies! @bdougie 

I'm working on it, No blockers!

Was a little occupied, conclusion shortly 🙏🏾🙏🏾

**Comment by brandonroberts** [comment]
@babblebey looks great! The modal doesn't close when I click on a user from the search results though

**Comment by babblebey** [comment]
> babblebey looks great! The modal doesn't close when I click on a user from the search results though 

Oh wow! 🤦🏾‍♂️

I can't believe I skipped that, and I considered that in my `***KeyboardCtrls` onClick `Enter` key.


I will address it, thanks for pointing this to me @brandonroberts

**Comment by bdougie** [comment]
Took a look on mobile and the logged out view is not quite there.

image

I also confirmed that I cannot select any users in the list. OnClick doesn't work but when I long press on iOS I can open via the peak screen.

image

**Comment by babblebey** [comment]
> Took a look on mobile and the logged out view is not quite there.

Yea yea! The Arrangement on that `TopNav` component really needs some refactoring! 

I had the SearchTrigger seat inside of one of the components specifically `AuthSection` to hold it in the `flex` there.

I'll appropriate.

> I also confirmed that I cannot select any users in the list. OnClick doesn't work but when I long press on iOS I can open via the peak screen.

It actually does work! It navigates to the User's profile behind the Modal! 

But, the Modal's `openSearch` state is still set to true 😅, exactly what @brandonroberts talked about

**Comment by babblebey** [comment]
### New Changes

- **Fix Top Nav:** I refactored the `TopNav` component extracting the `nav` element that wraps the menu items into a Separate component, and reused in a separate `div` optimized for mobile.
- **Move `openSearch` state:** I moved the `openSearch` state to the `GlobalState`'s `store`, this because the component is getting drilled down too many components within the `SearchDialog` and becoming uneasy to manage.
- **Implement Routing in `UserResultCard`:** I implemented `routing` into the `UserResultCard` in place of default `anchor` navigation action; I found that a click on the `anchor` element's navigation to the required route/page is noticeably slower than `routing`; this causes the `SearchDialog` to close (via `onClick` action first) but page still waits before loading the required route/page (I guess the speed varies depending on user's network connection). The usage of `routing` proves faster as its the functionality already implemented with selection via `KeyboardCtrls`.

```js
const handleClick: React.UIEventHandler<HTMLAnchorElement> = (e) => {
  e.preventDefault(); // <- Stops Original anchor/Link navigation action ❌ 
  router.push(e.currentTarget.href); // <- Uses Routing instead 🍕🍕🍕 
  setOpenSearch(false); // <- Close SearchDialog afterwards
};

<Link
  href={`/user/${login}`}
  ....
  onClick={handleClick}
>
  <Avatar size="sm" className="!rounded-full flex-none" avatarURL={getAvatarByUsername(login)} />
  <div className="flex items-center gap-2 overflow-hidden">
    .....
  </div>
</Link>
```
🍕

File: interfaces/global-state-types.ts [hunk 1]
```diff
@@ -10,4 +10,5 @@ export interface GlobalStateInterface {
   providerToken?: string | null;
   userId?: number | null;
   hasReports?: boolean;
+  openSearch?: boolean;
```
Comment (OgDev-01) [inline interfaces/global-state-types.ts:13]: Curious why this is global?

Comment (babblebey) [reply interfaces/global-state-types.ts:13]: Great question haha 😅

It was getting somewhat uneasy to manage the `openSearch` state, as I have started to drill down the `setOpenSearch` dispatch function almost 4 components deep to the search results list item.

State Tree

- AuthSection (Declares the `openSearch` state)
  - SearchDialogTrigger (uses the `setOpenSearch`)
  - SearchDialog (receives the `setOpenSearch` as prop from AuthSection to close the dialog onClick/keyboard shortcuts)
    - SearchResult (receives the `setOpenSearch` as prop from SearchDialog to pass it down to its child component`
      - UserResultCard - (receives the `setOpenSearch` as prop from `SearchResult` to close dialog onClick/keyboardCtrl result item.

Comment (OgDev-01) [reply interfaces/global-state-types.ts:13]: Nice! 👍. Looks good to me

**Review by OgDev-01 — APPROVED** [review]

File: lib/hooks/useDebounceTerm.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { useEffect, useRef, useState } from "react";
+
+const useDebounceTerm = (value: string, delay = 500) => {
+  const [debouncedValue, setDebouncedValue] = useState("");
+  const timeoutRef = useRef();
+
+  useEffect(() => {
+    timeoutRef.current = setTimeout(() => setDebouncedValue(value), delay) as unknown as typeof timeoutRef.current;
```
Comment (nickytonline) [inline lib/hooks/useDebounceTerm.ts:8]: Refactor (blocking): use the native type.

```suggestion
  const timeoutRef = useRef<NodeJS.Timeout>();

  useEffect(() => {
    timeoutRef.current = setTimeout(() => setDebouncedValue(value), delay);
```

File: components/organisms/TopNav/top-nav.tsx [hunk 3]
```diff
@@ -18,43 +33,30 @@ const TopNav: React.FC = () => {
...
-                  Highlights
-                </Link>
-              </li>
-            </ul>
-          </nav>
-        </div>
-        <AuthSection />
-      </div>
-    </header>
+    <nav className={className}>
```
Comment (nickytonline) [inline components/organisms/TopNav/top-nav.tsx:36]: Refactor (blocking): There are more than one `<nav />` elements on the page. To help screen readers, [add an aria-label for the nav element](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/navigation_role#associated_wai-aria_roles_states_and_properties). Insights seem to make sense for the name, but open to suggestions. See 

```suggestion
    <nav className={className} aria-label="Insights">
```

File: components/molecules/AuthSection/auth-section.tsx [hunk 4]
```diff
@@ -92,50 +96,56 @@ const AuthSection: React.FC = ({}) => {
...
+          <SearchDialogTrigger />
+          {user ? (
+            <>
+              {onboarded === false ? (
+                <>
+                  <OnboardingButton />
+                  <Divider type="vertical" className="!h-6 !bg-light-slate-6"></Divider>
+                </>
+              ) : (
+                ""
```
Comment (nickytonline) [inline components/molecules/AuthSection/auth-section.tsx:111]: Refactor: `null` is prefered when not returning anything in JSX.

```suggestion
                null
```

or you can do `!onboarded && your_JSX`

Comment (babblebey) [reply components/molecules/AuthSection/auth-section.tsx:111]: It was outside the scope of the PR at first!

But, I'll appropriate 🤞

File: components/organisms/TopNav/top-nav.tsx (hunk 3 above)
Comment (babblebey) [reply components/organisms/TopNav/top-nav.tsx:36]: Yea, its something around `main`

**Comment by babblebey** [comment]
🍕

**Review by brandonroberts — APPROVED** [review]
🍕

**Pull request merged**